### DISCUSSION TITLE
能否取消发送到设备时的路径翻译？

### DISCUSSION THREAD

**Discussion by shihong5858** [Cirn09/calibre-do-not-translate-my-path#25] [General]
其实是在用kobo，然后发现每次用发送到设备还是会给翻译成纯英文的路径。

**Comment by shihong5858** [comment]
![image](https://github.com/Cirn09/calibre-do-not-translate-my-path/assets/3605219/178c869b-74d3-4dd7-97af-08ecc7f43413)

**Comment by Cirn09** [comment]
### 是否能取消
理论上可行，但是很麻烦。比较合适的方法就是增加模板字段，但是本项目的实现方式是有局限的，可以简单地实现减少代码长度的功能，对增加代码长度的功能是非常困难的。很明显，添加字段是增加代码长度的修改

### 是否有必要取消
我认为没必要，大部分阅读器设备，至少我用过的 kindle、文石都支持显示文档标题而不是文件名，除非你发送的是 txt 这种纯文本

https://github.com/Cirn09/calibre-do-not-translate-my-path/issues/6

**Reply by shihong5858** [reply]
确实是，只是kobo支持自己拖拽文件，所以自己转好的文件（kobo要先用calibre转一次专用格式）会放到一个文件夹并且有可读的中文名称，但用send to Kindle传输过去的则完全是作者拼音乱序放到根目录下（类似此项目应用之前的电脑硬盘内状态）。

**Reply by shihong5858** [reply]
因为不清楚本项目的实现方式，所以仅仅是提出一个可能的需求~

希望大佬有空可以探索一下插件是否也能实现本项目的功能，如果能就实在是太爽了~
辛苦大佬回复~

**Reply by Cirn09** [reply]
不太对，好像不是增加字段，似乎是路径格式化之后有个拉丁化操作（未代码验证）
如果是这样的话，这样本项目可以实现这个功能，但是我缺少动力去实现这个（笑（甚至没动力去翻代码
功能多了排列组合搞出一堆 patch……

**Reply by shihong5858** [reply]
理解~没事的等什么时候觉得有需求了再说吧~

**Reply by KervynH** [reply]
Kindle的app显示的书籍名称是文件名。。。

**Reply by 381664** [reply]
附议，我是顺着网线爬来的。新买了文石notex3pro，发现就没有读取标题的功能，但是文石tab8c上就有。
坐等大佬更新，谢谢

**Comment by Cirn09** [comment]
> Kindle的app显示的书籍名称是文件名。。。

还有这事？太久没用 kindle 都忘了

**Reply by KervynH** [reply]
只能说Kindle确实有点是抽象
![Screenshot_20240322-122920](https://github.com/Cirn09/calibre-do-not-translate-my-path/assets/123544684/d2967da5-5a43-4430-906d-16542e966afb)

**Comment by shihong5858** [comment]
回报一下大佬，确实是有个拉丁化操作。
另外一个Kobo隔空传书的常用工具（https://github.com/daniel-j/send2ereader）也是一样的问题，打开看了js源代码也是先做了拉丁化操作。不太理解为什么做这个。

**Reply by HowieHz** [reply]
可能避免乱码啥的吧，但现在一般设备都支持 utf-8 文件名了

**Reply by HowieHz** [reply]
感觉可能是对日本人比较重要？毕竟那边用的编码有点怪。

**Comment by HowieHz** [comment]
之前自己写了个 fork，因为每次更新都要手动更新代码。看到这个项目之后，作为一个懒狗的我删掉了 fork。
参考 https://github.com/kovidgoyal/calibre/compare/master...kurikomoe:calibre-utf8-path:poc
和  https://github.com/snomiao/calibre-utf8-path

写好之后发送的文件是中文
![image](https://github.com/Cirn09/calibre-do-not-translate-my-path/assets/94725606/5855bc2f-c593-434c-b408-6161554f7149)

今天用这个项目的 calibre 发送的时候发现成拉丁文了。
看到这个讨论，感觉可以研究下差了啥没修改，应该不会太麻烦吧。。。

:: 这个项目代码是注入的方式而不是直接修改源代码，提 pr 还得研究下

**Reply by HowieHz** [reply]
找不出，放弃了哈哈

**Reply by HowieHz** [reply]
菜寄，
尝试从写入文件入手，没找到哪复制写入文件的函数。
尝试从按钮入手，太久没写 pyqt 了，
找到个 `a.triggered.connect(iactions['Send To Device'].menuless_qaction.trigger)` 就无从下手了
没找到 trigger

找老电脑回复了自己那个分支

src/calibre/ebooks/mobi/writer2/main.py
src/calibre/ebooks/mobi/writer8/mobi.py
中
`title = title.encode('ascii')`可能要改成`title = title.encode('utf-8')`
或者
```python
try:
  title = title.encode('ascii')
except:
  title = title.encode('utf-8')
```

没修改的话 mobi 转换可能会发生错误？需要测试下

另外看到 src/calibre/db/write.py
`book_id_val_map = {k:v for k, v in book_id_val_map.items() if v != g(k, missing)}` 改成了
```python
# FIXME(kuriko): force update all items whether changed or not
#  book_id_val_map = {k:v for k, v in book_id_val_map.items() if v != g(k, missing)}
book_id_val_map = {k:v for k, v in book_id_val_map.items()}
```

不知道和这个写入文件转换有没有关系

**Comment by HowieHz** [comment]
![image](https://github.com/Cirn09/calibre-do-not-translate-my-path/assets/94725606/7504c812-9a44-401d-830d-b71a28fdeb0c)

是不是应该直接注入 src\calibre\utils\filenames.py 的 ascii_filename，我检索了下发现很多直接
`from calibre.utils.filenames import ascii_filename`
就是下面这个函数

```python
def ascii_filename(orig, substitute='_'):
    if isinstance(substitute, bytes):
        substitute = substitute.decode(filesystem_encoding)
    orig = ascii_text(orig).replace('?', '_')
    ans = ''.join(x if ord(x) >= 32 else substitute for x in orig)
    return sanitize_file_name(ans, substitute=substitute)
```

我看 kuriko 大佬也是直接改 src/calibre/utils/filenames.py 的

```python
# FIXME(kuriko): use safe_filename to replace the ascii escape method
def safe_filename(filename):
    import re
    return re.sub(r"[\/\\\:\*\?\"\<\>\|]", "_", filename)

ascii_filename = safe_filename
```

仅仅 src\calibre\db\backend.py 就没啥效果了

**Reply by HowieHz** [reply]
折腾半天没折腾出来，注入也没注入明白

只能折中下了，直接保存到硬盘

![image](https://github.com/Cirn09/calibre-do-not-translate-my-path/assets/94725606/35b5e63e-08f0-4cc1-b393-cc887332b902)

**Reply by 381664** [reply]
传输到设备好像是通过插件完成的

**Comment by kilesduli** [comment]
我观察发现，calibre 自带了一个 `sanitize_file_name` 用来处理路径，于是我就直接把我需要的部分 `ascii_filename` 替换成了 `sanitize_file_name`。 一个就是 backend 的部分也就是用来处理 library folder的。另一个的就是发送到设备的时候路径翻译。其他的我就不管了。

其中 mtp 和 kobo 的部分我没有测试，usb 和无线连接是没有问题的。

这是 patch: 
https://gist.githubusercontent.com/kilesduli/0ab766c7b14a5a917b2f9fe0f932dc95
https://gist.githubusercontent.com/kilesduli/0ab766c7b14a5a917b2f9fe0f932dc95/raw/76ee30b39afc7b2b43d86454aae66d7409ffd47c/0001-feat-avoid-latinization-of-paths-in-certain-function.patch

使用方法：克隆 calibre 的仓库，应用 patch，然后自己编译。
或者，应用 patch 后，用这个仓库作者的工具 compile 一下修改的文件，生成 `python-lib.bypy.frozen` 然后自行替换。

**Reply by 381664** [reply]
厉害，我不会编程，我是顺着网线来的，我是个伸手党，哭

**Reply by kilesduli** [reply]
update 了一下，上次改完没测试，加的东西其实是有问题的，现在应该没问题了

**Comment by Cirn09** [comment]
可以了，快试试基于插件的 v3 版本吧
https://github.com/Cirn09/calibre-do-not-translate-my-path/releases/tag/v3.0.0

**Reply by HowieHz** [reply]
赞美作者

**Reply by 381664** [reply]
完美，感谢！
